File: Android Studio/JugnooNewUI/jugnooNewUI/src/main/java/product/clicklabs/jugnoo/promotion/fragments/ReferralsFragment.java
package product.clicklabs.jugnoo.promotion.fragments;

import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.os.Bundle;
import android.support.design.widget.TabLayout;
import android.support.v4.app.Fragment;
import android.support.v4.view.ViewPager;
import android.text.Spannable;
import android.text.SpannableString;
import android.text.SpannableStringBuilder;
import android.text.Spanned;
import android.text.method.LinkMovementMethod;
import android.text.style.ClickableSpan;
import android.text.style.RelativeSizeSpan;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.RelativeLayout;
import android.widget.TextView;

import com.sabkuchfresh.analytics.GAAction;
import com.sabkuchfresh.analytics.GACategory;
import com.sabkuchfresh.analytics.GAUtils;
import com.squareup.picasso.Picasso;

import product.clicklabs.jugnoo.Data;
import product.clicklabs.jugnoo.MyApplication;
import product.clicklabs.jugnoo.R;
import product.clicklabs.jugnoo.promotion.ReferralActions;
import product.clicklabs.jugnoo.promotion.ShareActivity;
import product.clicklabs.jugnoo.promotion.adapters.MediaInfoFragmentAdapter;
import product.clicklabs.jugnoo.promotion.dialogs.ReferDriverDialog;
import product.clicklabs.jugnoo.utils.ASSL;
import product.clicklabs.jugnoo.utils.DialogPopup;
import product.clicklabs.jugnoo.utils.Fonts;
import product.clicklabs.jugnoo.utils.Utils;


public class ReferralsFragment extends Fragment implements  GACategory, GAAction{

	private RelativeLayout relativeLayoutRoot;

	private ImageView imageViewLogo;
	private TextView textViewDesc, textViewLeaderboardSingle;
	private RelativeLayout relativeLayoutReferContainer, relativeLayoutReferSingle, relativeLayoutMultipleTab, relativeLayoutLeaderboardSingle;
	private LinearLayout linearLayoutLeaderBoard, linearLayoutRefer;
	private View rootView;
    private ShareActivity activity;
    private Bundle bundle;

	private LinearLayout llWhatsappShare, llReferralCode;
	private TextView tvMoreSharingOptions, tvReferralCodeValue, textViewReferDriver;
	private LinearLayout llUserReferralData;
	private TextView tvReferralsCount, tvCashEarned, tvCashEarnedToday;
	private RelativeLayout rlViewPager;
	private ViewPager viewPagerImageVideo;
	private MediaInfoFragmentAdapter mediaInfoFragmentAdapter;
	private TabLayout tabDots;


	public static ReferralsFragment newInstance(){
		ReferralsFragment fragment = new ReferralsFragment();
		Bundle bundle = new Bundle();
		fragment.setArguments(bundle);
		return fragment;
	}


    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        rootView = inflater.inflate(R.layout.fragment_referrals, container, false);


        activity = (ShareActivity) getActivity();
        bundle = new Bundle();
		relativeLayoutRoot = (RelativeLayout) rootView.findViewById(R.id.relativeLayoutRoot);
		try {
			if(relativeLayoutRoot != null) {
				new ASSL(activity, relativeLayoutRoot, 1134, 720, false);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}


		GAUtils.trackScreenView(REFERRAL+HOME);

		imageViewLogo = (ImageView) rootView.findViewById(R.id.imageViewLogo);
		textViewDesc = (TextView)rootView.findViewById(R.id.textViewDesc);textViewDesc.setTypeface(Fonts.mavenMedium(activity));

		textViewLeaderboardSingle = (TextView)rootView.findViewById(R.id.textViewLeaderboardSingle);textViewLeaderboardSingle.setTypeface(Fonts.mavenMedium(activity));

		relativeLayoutReferContainer = (RelativeLayout) rootView.findViewById(R.id.relativeLayoutReferContainer);
		relativeLayoutReferSingle = (RelativeLayout) rootView.findViewById(R.id.relativeLayoutReferSingle);
		relativeLayoutMultipleTab = (RelativeLayout) rootView.findViewById(R.id.relativeLayoutMultipleTab);
		relativeLayoutLeaderboardSingle = (RelativeLayout) rootView.findViewById(R.id.relativeLayoutLeaderboardSingle);
		linearLayoutLeaderBoard = (LinearLayout) rootView.findViewById(R.id.linearLayoutLeaderBoard);
		((TextView)rootView.findViewById(R.id.textViewLeaderboard)).setTypeface(Fonts.mavenMedium(activity));
		linearLayoutRefer = (LinearLayout)rootView.findViewById(R.id.linearLayoutRefer);
		textViewReferDriver = rootView.findViewById(R.id.textViewReferDriver);
		textViewReferDriver.setTypeface(Fonts.mavenMedium(activity));

		llUserReferralData = rootView.findViewById(R.id.llUserReferralData);
		tvReferralsCount = rootView.findViewById(R.id.tvReferralsCount);
		tvCashEarned = rootView.findViewById(R.id.tvCashEarned);
		tvCashEarnedToday = rootView.findViewById(R.id.tvCashEarnedToday);
		rlViewPager = rootView.findViewById(R.id.rlViewPager);
		viewPagerImageVideo = rootView.findViewById(R.id.viewPagerImageVideo);
		tabDots = rootView.findViewById(R.id.tabDots);


		llWhatsappShare = (LinearLayout) rootView.findViewById(R.id.llWhatsappShare);
		llReferralCode = (LinearLayout) rootView.findViewById(R.id.llReferralCode);
		((TextView) rootView.findViewById(R.id.tvWhatsapp)).setTypeface(Fonts.mavenMedium(activity));
		tvMoreSharingOptions = (TextView) rootView.findViewById(R.id.tvMoreSharingOptions);
		tvMoreSharingOptions.setTypeface(Fonts.mavenRegular(activity));
		Utils.setTextUnderline(tvMoreSharingOptions, activity.getString(R.string.view_more_sharing_options));
		((TextView)rootView.findViewById(R.id.tvYourReferralCode)).setTypeface(Fonts.mavenRegular(activity));
		tvReferralCodeValue = (TextView) rootView.findViewById(R.id.tvReferralCodeValue);
		tvReferralCodeValue.setTypeface(Fonts.mavenMedium(activity));


		tvMoreSharingOptions.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View view) {
				if(MyApplication.getInstance().isOnline()) {
					ReferralActions.openGenericShareIntent(activity, activity.getCallbackManager());
					try {
						if(activity.fromDeepLink){
						} else{
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
					GAUtils.event(SIDE_MENU, FREE_GIFT, MORE_SHARING_OPTIONS+CLICKED);
				} else{
					DialogPopup.alertPopup(activity, "", activity.getString(R.string.connection_lost_desc));
				}
			}
		});


		llWhatsappShare.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				if(Utils.appInstalledOrNot(activity, "com.whatsapp")){
					ReferralActions.shareToWhatsapp(activity);
				} else {
					ReferralActions.openGenericShareIntent(activity, null);
				}
				GAUtils.event(SIDE_MENU, FREE_GIFT, GAAction.WHATSAPP+INVITE+CLICKED);
			}
		});

		llReferralCode.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				try {
					ClipboardManager clipboard = (ClipboardManager) activity.getSystemService(Context.CLIPBOARD_SERVICE);
					ClipData clip = ClipData.newPlainText(tvReferralCodeValue.getText().toString(),
							tvReferralCodeValue.getText().toString());
					clipboard.setPrimaryClip(clip);
					Utils.showToast(activity, activity.getString(R.string.referral_code_copied));
					GAUtils.event(SIDE_MENU, FREE_GIFT, GAAction.REFERRAL+CODE_COPIED);
				} catch (Exception e) {}
			}
		});

		linearLayoutLeaderBoard.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				activity.openLeaderboardFragment();
			}
		});

		linearLayoutRefer.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				relativeLayoutReferSingle.performClick();
			}
		});

		relativeLayoutLeaderboardSingle.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				linearLayoutLeaderBoard.performClick();
			}
		});

		textViewReferDriver.setOnClickListener(new View.OnClickListener() {
			@Override
			public void onClick(View v) {
				getReferDriverDialog().show();
			}
		});

		tvReferralsCount.setOnClickListener(v -> activity.openReferralTxnFragment(ReferralTxnFragment.STATE_REFERRALS));
		tvCashEarned.setOnClickListener(v -> activity.openReferralTxnFragment(ReferralTxnFragment.STATE_TOTAL));
		tvCashEarnedToday.setOnClickListener(v -> activity.openReferralTxnFragment(ReferralTxnFragment.STATE_TODAY));

		try {

			if(Data.userData != null && Data.userData.getReferralMessages().getMultiLevelReferralEnabled()){
				llUserReferralData.setVisibility(View.VISIBLE);

				setHighlightText(tvReferralsCount, getString(R.string.referrals), String.valueOf((int)Data.userData.getReferralMessages().getReferralsCount()));
				setHighlightText(tvCashEarned, getString(R.string.cash_earned), Utils.formatCurrencyValue(Data.autoData.getCurrency(), Data.userData.getReferralMessages().getReferralEarnedTotal()));
				setHighlightText(tvCashEarnedToday, getString(R.string.earned_today), Utils.formatCurrencyValue(Data.autoData.getCurrency(), Data.userData.getReferralMessages().getReferralEarnedToday()));



				if(Data.userData.getReferralMessages().getReferralImages() == null
						|| Data.userData.getReferralMessages().getReferralImages().size() == 0){  // images length 0
					rlViewPager.setVisibility(View.GONE);
					imageViewLogo.setVisibility(View.VISIBLE);
				} else {
					rlViewPager.setVisibility(View.VISIBLE);
					imageViewLogo.setVisibility(View.GONE);

					mediaInfoFragmentAdapter = new MediaInfoFragmentAdapter(getChildFragmentManager(), Data.userData.getReferralMessages().getReferralImages());
					viewPagerImageVideo.setAdapter(mediaInfoFragmentAdapter);
					tabDots.setupWithViewPager(viewPagerImageVideo);

					setTabLayoutMargin(tabDots);
				}

			} else {
				llUserReferralData.setVisibility(View.GONE);
				rlViewPager.setVisibility(View.GONE);
				imageViewLogo.setVisibility(View.VISIBLE);
			}


			tvReferralCodeValue.setText(Data.userData.referralCode);
			textViewDesc.setText(Data.userData.getReferralMessages().referralShortMessage+" "+getString(R.string.details));

			SpannableString ss = new SpannableString(Data.userData.getReferralMessages().referralShortMessage+"\n "+getString(R.string.details));
			ClickableSpan clickableSpan = new ClickableSpan() {
				@Override
				public void onClick(View textView) {
					try {
						boolean html = Utils.DetectHtml.isHtml(Data.userData.getReferralMessages().referralMoreInfoMessage);
						DialogPopup.alertPopupLeftOriented(activity, "",
								Data.userData.getReferralMessages().referralMoreInfoMessage, true, true, html);
						GAUtils.event(SIDE_MENU, FREE_GIFT, GAAction.DETAILS+CLICKED);
					} catch (Exception e) {
						e.printStackTrace();
					}
				}
			};
			ss.setSpan(clickableSpan, (textViewDesc.getText().length() - 6), textViewDesc.getText().length() + 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE);

			textViewDesc.setText(ss);
			textViewDesc.setMovementMethod(LinkMovementMethod.getInstance());

			if(!"".equalsIgnoreCase(Data.userData.getInviteEarnScreenImage())){
				Picasso.with(activity).load(Data.userData.getInviteEarnScreenImage())
						.into(imageViewLogo);
			}

			relativeLayoutReferContainer.setVisibility(View.GONE);
			if (Data.userData != null) {
				relativeLayoutMultipleTab.setVisibility(View.GONE);
				relativeLayoutReferSingle.setVisibility(View.GONE);
				relativeLayoutLeaderboardSingle.setVisibility(View.GONE);
				if (Data.userData.getReferralLeaderboardEnabled() == 1 && Data.userData.getcToDReferralEnabled() == 1) {
					relativeLayoutMultipleTab.setVisibility(View.VISIBLE);
					relativeLayoutReferContainer.setVisibility(View.VISIBLE);
				}
				else if (Data.userData.getReferralLeaderboardEnabled() == 1 && Data.userData.getcToDReferralEnabled() != 1) {
					relativeLayoutLeaderboardSingle.setVisibility(View.VISIBLE);
					relativeLayoutReferContainer.setVisibility(View.VISIBLE);
				}
				else if (Data.userData.getReferralLeaderboardEnabled() != 1 && Data.userData.getcToDReferralEnabled() == 1) {
					relativeLayoutReferSingle.setVisibility(View.VISIBLE);
					relativeLayoutReferContainer.setVisibility(View.VISIBLE);
				}
			}

		} catch(Exception e){
			e.printStackTrace();
		}


		return rootView;
	}

	private ReferDriverDialog referDriverDialog;
	private ReferDriverDialog getReferDriverDialog(){
		if(referDriverDialog == null){
			referDriverDialog = new ReferDriverDialog(activity);
		}
		return referDriverDialog;
	}


    @Override
	public void onDestroy() {
		super.onDestroy();
        ASSL.closeActivity(relativeLayoutRoot);
        System.gc();
	}

	private void setHighlightText(TextView textView, String normal, String highlight){
		textView.setText(normal);
		SpannableStringBuilder ssb = new SpannableStringBuilder(highlight);
		ssb.setSpan(new RelativeSizeSpan(1f), 0, ssb.length(), Spannable.SPAN_EXCLUSIVE_EXCLUSIVE);

		textView.append("\n");
		textView.append(ssb);
	}

	private void setTabLayoutMargin(TabLayout tabDots){
		if (tabDots.getTabCount() == 1) {
			tabDots.setVisibility(View.GONE);
		} else {
			tabDots.setVisibility(View.VISIBLE);
			for (int i = 0; i < tabDots.getTabCount(); i++) {
				View tab = ((ViewGroup) tabDots.getChildAt(0)).getChildAt(i);
				ViewGroup.MarginLayoutParams p = (ViewGroup.MarginLayoutParams) tab.getLayoutParams();
				p.setMargins(activity.getResources().getDimensionPixelSize(R.dimen.dp_4), 0, 0, 0);
				p.setMarginStart(activity.getResources().getDimensionPixelSize(R.dimen.dp_4));
				p.setMarginEnd(0);
				tab.requestLayout();
			}
		}
	}

}

File: Android Studio/JugnooNewUI/jugnooNewUI/src/main/java/product/clicklabs/jugnoo/promotion/fragments/ReferralsFragment.kt
package product.clicklabs.jugnoo.promotion.fragments

import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.os.Bundle
import com.google.android.material.tabs.TabLayout
import androidx.fragment.app.Fragment
import android.text.Spannable
import android.text.SpannableString
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.method.LinkMovementMethod
import android.text.style.ClickableSpan
import android.text.style.RelativeSizeSpan
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import com.sabkuchfresh.analytics.GAAction
import com.sabkuchfresh.analytics.GACategory
import com.sabkuchfresh.analytics.GAUtils
import com.sabkuchfresh.feed.ui.api.APICommonCallback
import com.sabkuchfresh.feed.ui.api.ApiCommon
import com.sabkuchfresh.feed.ui.api.ApiName
import com.squareup.picasso.Picasso
import kotlinx.android.synthetic.main.fragment_referrals.*
import product.clicklabs.jugnoo.Data
import product.clicklabs.jugnoo.MyApplication
import product.clicklabs.jugnoo.R
import product.clicklabs.jugnoo.home.HomeActivity
import product.clicklabs.jugnoo.promotion.ReferralActions
import product.clicklabs.jugnoo.promotion.ShareActivity
import product.clicklabs.jugnoo.promotion.adapters.MediaInfoFragmentAdapter
import product.clicklabs.jugnoo.promotion.dialogs.ReferDriverDialog
import product.clicklabs.jugnoo.promotion.models.ReferralTxnResponse
import product.clicklabs.jugnoo.utils.DateOperations
import product.clicklabs.jugnoo.utils.DialogPopup
import product.clicklabs.jugnoo.utils.Fonts
import product.clicklabs.jugnoo.utils.Utils
import java.util.*


class ReferralsFragment : Fragment(), GACategory, GAAction {


    private var rootView: View? = null
    private var activity: ShareActivity? = null
    private var bundle: Bundle? = null

    private var mediaInfoFragmentAdapter: MediaInfoFragmentAdapter? = null

    private var referDriverDialog: ReferDriverDialog? = null

    private var referralTxnResponse: ReferralTxnResponse? = null


    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        rootView = inflater.inflate(R.layout.fragment_referrals, container, false)
        return rootView
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        activity = requireActivity() as ShareActivity
        bundle = Bundle()


        GAUtils.trackScreenView(GAAction.REFERRAL + GAAction.HOME)

        textViewDesc!!.typeface = Fonts.mavenMedium(activity)

        textViewLeaderboardSingle!!.typeface = Fonts.mavenMedium(activity)

        textViewLeaderboard.typeface = Fonts.mavenMedium(activity)
        textViewReferDriver!!.typeface = Fonts.mavenMedium(activity)



        tvWhatsapp.typeface = Fonts.mavenMedium(activity)
        tvMoreSharingOptions!!.typeface = Fonts.mavenRegular(activity)
        Utils.setTextUnderline(tvMoreSharingOptions, activity!!.getString(R.string.view_more_sharing_options))
        tvYourReferralCode.typeface = Fonts.mavenRegular(activity)
        tvReferralCodeValue!!.typeface = Fonts.mavenMedium(activity)


        tvMoreSharingOptions!!.setOnClickListener {
            if (MyApplication.getInstance().isOnline) {
                ReferralActions.openGenericShareIntent(activity, activity!!.callbackManager)

                GAUtils.event(GACategory.SIDE_MENU, GAAction.FREE_GIFT, GAAction.MORE_SHARING_OPTIONS + GAAction.CLICKED)
            } else {
                DialogPopup.alertPopup(activity, "", activity!!.getString(R.string.connection_lost_desc))
            }
        }


        llWhatsappShare!!.setOnClickListener {
            if (Utils.appInstalledOrNot(activity!!, "com.whatsapp")) {
                ReferralActions.shareToWhatsapp(activity)
            } else {
                ReferralActions.openGenericShareIntent(activity, null)
            }
            GAUtils.event(GACategory.SIDE_MENU, GAAction.FREE_GIFT, GAAction.WHATSAPP + GAAction.INVITE + GAAction.CLICKED)
        }

        llReferralCode!!.setOnClickListener {
            try {
                val clipboard = activity!!.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
                val clip = ClipData.newPlainText(tvReferralCodeValue!!.text.toString(),
                        tvReferralCodeValue!!.text.toString())
                clipboard.setPrimaryClip(clip)
                Utils.showToast(activity, activity!!.getString(R.string.referral_code_copied))
                GAUtils.event(GACategory.SIDE_MENU, GAAction.FREE_GIFT, GAAction.REFERRAL + GAAction.CODE_COPIED)
            } catch (e: Exception) {
            }
        }

        linearLayoutLeaderBoard!!.setOnClickListener { activity!!.openLeaderboardFragment() }

        linearLayoutRefer!!.setOnClickListener { relativeLayoutReferSingle!!.performClick() }

        relativeLayoutLeaderboardSingle!!.setOnClickListener { linearLayoutLeaderBoard!!.performClick() }

        textViewReferDriver!!.setOnClickListener { getReferDriverDialog().show() }

        tvReferralsCount!!.setOnClickListener { v -> openReferralTxnFragment(ReferralTxnFragment.STATE_REFERRALS) }
        tvCashEarned!!.setOnClickListener { v -> openReferralTxnFragment(ReferralTxnFragment.STATE_TOTAL) }
        tvCashEarnedToday!!.setOnClickListener { v -> openReferralTxnFragment(ReferralTxnFragment.STATE_TODAY) }

        try {

            if (Data.userData != null && Data.userData.referralMessages.multiLevelReferralEnabled) {
                referralInfoApi()

                llUserReferralData!!.visibility = View.VISIBLE

                setHighlightText(tvReferralsCount!!, getString(R.string.referrals), Data.userData.referralMessages.referralsCount.toInt().toString())
                setHighlightText(tvCashEarned!!, getString(R.string.cash_earned), Utils.formatCurrencyValue(Data.autoData.currency, Data.userData.referralMessages.referralEarnedTotal))
                setHighlightText(tvCashEarnedToday!!, getString(R.string.earned_today), Utils.formatCurrencyValue(Data.autoData.currency, Data.userData.referralMessages.referralEarnedToday))



                if (Data.userData.referralMessages.referralImages == null || Data.userData.referralMessages.referralImages.size == 0) {  // images length 0
                    rlViewPager!!.visibility = View.GONE
                    imageViewLogo!!.visibility = View.VISIBLE
                } else {
                    rlViewPager!!.visibility = View.VISIBLE
                    imageViewLogo!!.visibility = View.GONE

                    mediaInfoFragmentAdapter = MediaInfoFragmentAdapter(childFragmentManager, Data.userData.referralMessages.referralImages)
                    viewPagerImageVideo!!.adapter = mediaInfoFragmentAdapter
                    tabDots!!.setupWithViewPager(viewPagerImageVideo)

                    setTabLayoutMargin(tabDots!!)
                }

            } else {
                llUserReferralData!!.visibility = View.GONE
                rlViewPager!!.visibility = View.GONE
                imageViewLogo!!.visibility = View.VISIBLE
            }


            tvReferralCodeValue!!.text = Data.userData.referralCode
            textViewDesc!!.text = Data.userData.referralMessages.referralShortMessage + " " + getString(R.string.details)

            val ss = SpannableString(Data.userData.referralMessages.referralShortMessage + "\n " + getString(R.string.details))
            val clickableSpan = object : ClickableSpan() {
                override fun onClick(textView: View) {
                    try {
                        val html = Utils.DetectHtml.isHtml(Data.userData.referralMessages.referralMoreInfoMessage)
                        DialogPopup.alertPopupLeftOriented(activity, "",
                                Data.userData.referralMessages.referralMoreInfoMessage, true, true, html)
                        GAUtils.event(GACategory.SIDE_MENU, GAAction.FREE_GIFT, GAAction.DETAILS + GAAction.CLICKED)
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }

                }
            }
            ss.setSpan(clickableSpan, textViewDesc!!.text.length - getString(R.string.details).length, textViewDesc!!.text.length + 1, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)

            textViewDesc!!.text = ss
            textViewDesc!!.movementMethod = LinkMovementMethod.getInstance()

            if (!"".equals(Data.userData.inviteEarnScreenImage, ignoreCase = true)) {
                Picasso.with(activity).load(Data.userData.inviteEarnScreenImage)
                        .into(imageViewLogo)
            }

            relativeLayoutReferContainer!!.visibility = View.GONE
            if (Data.userData != null) {
                relativeLayoutMultipleTab!!.visibility = View.GONE
                relativeLayoutReferSingle!!.visibility = View.GONE
                relativeLayoutLeaderboardSingle!!.visibility = View.GONE
                if (Data.userData.referralLeaderboardEnabled == 1 && Data.userData.getcToDReferralEnabled() == 1) {
                    relativeLayoutMultipleTab!!.visibility = View.VISIBLE
                    relativeLayoutReferContainer!!.visibility = View.VISIBLE
                } else if (Data.userData.referralLeaderboardEnabled == 1 && Data.userData.getcToDReferralEnabled() != 1) {
                    relativeLayoutLeaderboardSingle!!.visibility = View.VISIBLE
                    relativeLayoutReferContainer!!.visibility = View.VISIBLE
                } else if (Data.userData.referralLeaderboardEnabled != 1 && Data.userData.getcToDReferralEnabled() == 1) {
                    relativeLayoutReferSingle!!.visibility = View.VISIBLE
                    relativeLayoutReferContainer!!.visibility = View.VISIBLE
                }
            }

        } catch (e: Exception) {
            e.printStackTrace()
        }


    }

    private fun setReferralStats(referralTxnResponse: ReferralTxnResponse?) {
        if(referralTxnResponse?.referralData != null
                && referralTxnResponse.referralData!!.referralsCount != null
                && referralTxnResponse.referralData!!.referralEarnedTotal != null
                && referralTxnResponse.referralData!!.referralEarnedToday != null) {
            setHighlightText(tvReferralsCount!!, getString(R.string.referrals),
                    referralTxnResponse.referralData!!.referralsCount!!.toInt().toString())
            setHighlightText(tvCashEarned!!, getString(R.string.cash_earned),
                    Utils.formatCurrencyValue(Data.autoData.currency, referralTxnResponse.referralData!!.referralEarnedTotal!!))
            setHighlightText(tvCashEarnedToday!!, getString(R.string.earned_today),
                    Utils.formatCurrencyValue(Data.autoData.currency, referralTxnResponse.referralData!!.referralEarnedToday!!))

            Data.userData.referralMessages.referralsCount = referralTxnResponse.referralData!!.referralsCount!!
            Data.userData.referralMessages.referralEarnedTotal = referralTxnResponse.referralData!!.referralEarnedTotal!!
            Data.userData.referralMessages.referralEarnedToday = referralTxnResponse.referralData!!.referralEarnedToday!!
        }
    }

    private fun getReferDriverDialog(): ReferDriverDialog {
        if (referDriverDialog == null) {
            referDriverDialog = ReferDriverDialog(activity)
        }
        return referDriverDialog!!
    }


    private fun setHighlightText(textView: TextView, normal: String, highlight: String) {
        textView.text = normal
        val ssb = SpannableStringBuilder(highlight)
        ssb.setSpan(RelativeSizeSpan(1f), 0, ssb.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)

        textView.append("\n")
        textView.append(ssb)
    }

    private fun setTabLayoutMargin(tabDots: TabLayout) {
        if (tabDots.tabCount == 1) {
            tabDots.visibility = View.GONE
        } else {
            tabDots.visibility = View.VISIBLE
            for (i in 0 until tabDots.tabCount) {
                val tab = (tabDots.getChildAt(0) as ViewGroup).getChildAt(i)
                val p = tab.layoutParams as ViewGroup.MarginLayoutParams
                p.setMargins(activity!!.resources.getDimensionPixelSize(R.dimen.dp_4), 0, 0, 0)
                p.marginStart = activity!!.resources.getDimensionPixelSize(R.dimen.dp_4)
                p.marginEnd = 0
                tab.requestLayout()
            }
        }
    }

    companion object {

        @JvmStatic
        fun newInstance(): ReferralsFragment {
            val fragment = ReferralsFragment()
            val bundle = Bundle()
            fragment.arguments = bundle
            return fragment
        }
    }


    private fun referralInfoApi(){

        if (!HomeActivity.checkIfUserDataNull(activity) && Data.userData.referralMessages.multiLevelReferralEnabled) {
            val params = HashMap<String, String>()
            ApiCommon<ReferralTxnResponse>(requireActivity()).showLoader(true)
                    .execute(params, ApiName.REFERRAL_INFO, object: APICommonCallback<ReferralTxnResponse>(){
                        override fun onSuccess(t: ReferralTxnResponse?, message: String?, flag: Int) {
                            referralTxnResponse = t
                            setReferralStats(referralTxnResponse)
                        }

                        override fun onError(t: ReferralTxnResponse?, message: String?, flag: Int): Boolean {
                            return false
                        }
                    })
        }

    }

    private fun openReferralTxnFragment(state:Int){
        activity!!.openReferralTxnFragment(state)
    }

    public fun getTxnList(state:Int):MutableList<Any>?{
        val list = mutableListOf<Any>()
        if(state == ReferralTxnFragment.STATE_REFERRALS
                && referralTxnResponse != null && referralTxnResponse!!.referralData != null && referralTxnResponse!!.referralData!!.referrals != null){
            list.addAll(referralTxnResponse!!.referralData!!.referrals!!)
        }
        else if(state == ReferralTxnFragment.STATE_TOTAL
                && referralTxnResponse != null && referralTxnResponse!!.referralData != null && referralTxnResponse!!.referralData!!.txns != null){
            list.addAll(referralTxnResponse!!.referralData!!.txns!!)
        }
        else if(state == ReferralTxnFragment.STATE_TODAY
                && referralTxnResponse != null && referralTxnResponse!!.referralData != null && referralTxnResponse!!.referralData!!.txns != null){
            val currentDate = DateOperations.getCurrentDate()
            val currentDateItems = referralTxnResponse!!.referralData!!.txns!!.filter {
                if(it.isToday != null){
                    it.isToday == 1
                } else {
                    it.creditedOn?.contains(currentDate) ?: false
                }
            }
            list.addAll(currentDateItems)
        }

        return list
    }

}
